xds/internal/xdsclient/transport/lrs/lrs_stream.go

```diff
@@ -0,0 +1,339 @@
+		if err != nil {
+			lrs.logger.Warningf("Creating new LRS stream failed: %v", err)
+			return nil
+		}
+		if lrs.logger.V(2) {
+			lrs.logger.Infof("LRS stream created")
+		}
+
+		if err := lrs.sendFirstLoadStatsRequest(stream, node); err != nil {
+			lrs.logger.Warningf("Sending first LRS request failed: %v", err)
+			return nil
+		}
+
+		clusters, interval, err := lrs.recvFirstLoadStatsResponse(stream)
+		if err != nil {
+			lrs.logger.Warningf("Reading from LRS stream failed: %v", err)
+			return nil
+		}
+
+		// We reset backoff state when we successfully receive at least one
+		// message from the server.
+		lrs.sendLoads(streamCtx, stream, clusters, interval)
+		return backoff.ErrResetBackoff
+	}
+	backoff.RunF(ctx, runLoadReportStream, lrs.backoff)
+}
+
+// sendLoads is responsible for periodically sending load reports to the LRS
+// server at the specified interval for the specified clusters, until the passed
+// in context is canceled.
+func (lrs *Stream) sendLoads(ctx context.Context, stream transport.StreamingCall, clusterNames []string, interval time.Duration) {
+	tick := time.NewTicker(interval)
+	defer tick.Stop()
+	for {
+		select {
+		case <-tick.C:
+		case <-ctx.Done():
+			return
+		}
+		if err := lrs.sendLoadStatsRequest(stream, lrs.lrsStore.Stats(clusterNames)); err != nil {
+			lrs.logger.Warningf("Writing to LRS stream failed: %v", err)
+			return
+		}
+	}
+}
+
+func (lrs *Stream) sendFirstLoadStatsRequest(stream transport.StreamingCall, node *v3corepb.Node) error {
+	req := &v3lrspb.LoadStatsRequest{Node: node}
+	if lrs.logger.V(perRPCVerbosityLevel) {
+		lrs.logger.Infof("Sending initial LoadStatsRequest: %s", pretty.ToJSON(req))
+	}
+	err := stream.Send(req)
+	if err == io.EOF {
+		return getStreamError(stream)
+	}
+	return err
+}
+
+// recvFirstLoadStatsResponse receives the first LoadStatsResponse from the LRS
+// server.  Returns the following:
+//   - a list of cluster names requested by the server or an empty slice if the
+//     server requested for load from all clusters
+//   - the load reporting interval, and
+//   - any error encountered
+//
+// If the server requests for endpoint-level load reporting, an error is
+// returned, since this is not yet supported.
+func (lrs *Stream) recvFirstLoadStatsResponse(stream transport.StreamingCall) ([]string, time.Duration, error) {
+	r, err := stream.Recv()
+	if err != nil {
+		return nil, 0, fmt.Errorf("lrs: failed to receive first LoadStatsResponse: %v", err)
+	}
+	resp, ok := r.(*v3lrspb.LoadStatsResponse)
+	if !ok {
+		return nil, time.Duration(0), fmt.Errorf("lrs: unexpected message type %T", r)
+	}
+	if lrs.logger.V(perRPCVerbosityLevel) {
+		lrs.logger.Infof("Received first LoadStatsResponse: %s", pretty.ToJSON(resp))
+	}
+
+	rInterval := resp.GetLoadReportingInterval()
+	if rInterval.CheckValid() != nil {
+		return nil, 0, fmt.Errorf("lrs: invalid load_reporting_interval: %v", err)
+	}
+	interval := rInterval.AsDuration()
```

Comment:
I always wonder what to do here when I get a variable from a methodA(), and then typecast it to data I'll use eventually which is semantically the same thing.

Reply:
I don't understand your comment. Do you want me to get rid of the local `interval` and inline `rInterval.AsDuration()` in the return statement?

Reply:
Ah I see. I don't think there is any specific guidance around this in the style guide, or at least, I haven't seen it before.

Personally, if I'm doing
```golang
data := getData()
data2, ok := data.(Data)
```

I would make the variable name for `data` as small as possible, since its scope is only a couple of lines. And I would make the variable name for `data2` to be more meaningful, since that would probably have a bigger scope. I had left the names as they were in the previous code, but changed it now to be more descriptive for the second one. I couldn't use `int` for interval since that is a reserved name, and I didn't want to use `i` for interval since that is usually reserved for indices.